src/convert_hl7v2_fhir/controllers/er7/er7_extractor.py

```diff
@@ -0,0 +1,115 @@
+from datetime import datetime, timezone
+from typing import List
+import hl7apy
+from hl7apy.core import Message
+
+from convert_hl7v2_fhir.controllers.er7.exceptions import (
+    MissingNHSNumberError,
+    InvalidNHSNumberError,
+    MissingPointOfCareError,
+    MissingFacilityError,
+    MissingPatientClassError,
+    MissingAdmissionTypeError,
+    MissingTimeOfAdmissionError,
+    MissingFamilyNameError,
+    MissingGivenNameError,
+)
+from convert_hl7v2_fhir.controllers.utils import is_nhs_number_valid
+
+
+class ER7Extractor:
+    def __init__(self, er7_message: Message):
+        self.er7_message = er7_message
+
+    def nhs_number(self) -> str:
+        """There are two potential fields where NHS number can be:
+        1. PID.2 - Patient Id
+        2. PID.3 - Patient Identifier List"""
+
+        nhs_numbers = [
+            composite_id.id.value
+            for composite_id in self.er7_message.pid.patient_id_internal_id
+            if "NHSNMBR" in composite_id.value
+        ]
+        if "NHSNMBR" in self.er7_message.pid.patient_id_external_id.value:
+            nhs_numbers.append(self.er7_message.pid.patient_id_external_id.id.value)
+
+        if not nhs_numbers:
+            raise MissingNHSNumberError
+
+        try:
+            return next(
+                (
+                    nhs_number
+                    for nhs_number in nhs_numbers
+                    if is_nhs_number_valid(nhs_number)
+                )
+            )
+        except StopIteration:
+            raise InvalidNHSNumberError
+
+    def family_name(self) -> str:
+        _family_name = self.er7_message.pid.patient_name.family_name.value
+        if not _family_name:
+            raise MissingFamilyNameError
+
+        return self.er7_message.pid.patient_name.family_name.value
+
+    def given_name(self) -> List[str]:
```

Comment:
```suggestion
    def given_names(self) -> List[str]:
```

Reply:
I get your point but I wanted the whole codebase to be in sync, and in fhir structure "given_name" (singular) field is a list...

Reply:
Ah, ok.  Leave as is in that case.  If I feel suitably enraged by it later I'll add an alias method.